test/migrator_test.rb

```diff
@@ -6,6 +6,62 @@
 
   before { ActiveRecord::Base.establish_connection(RAILS_ENV.to_sym) }
 
+  describe "when DB is empty" do
+    extend RailsEnvSwitch
+
+    switch_rails_env('test3')
+
+    it "makes meta tables" do
+      ActiveRecord::Base.on_shard(nil) do
+        refute table_exists?(:unsharded_table)
+        if ActiveRecord::VERSION::MAJOR >= 4
+          refute ActiveRecord::SchemaMigration.table_exists?
+        else
+          refute table_exists?(:schema_migrations)
+        end
+      end
+
+      ActiveRecord::Base.on_all_shards do
+        refute table_exists?(:sharded_table)
+        if ActiveRecord::VERSION::MAJOR >= 4
+          refute ActiveRecord::SchemaMigration.table_exists?
+        else
+          refute table_exists?(:schema_migrations)
+        end
+      end
+
+      migrator(:up, 'separate_migrations').migrate
+
+      ActiveRecord::Base.on_shard(nil) do
+        assert table_exists?(:unsharded_table)
+        if ActiveRecord::VERSION::MAJOR >= 4
+          assert ActiveRecord::SchemaMigration.table_exists?
+        else
+          assert table_exists?(:schema_migrations)
```

Comment:
I would still use `table_exists?` helper for Rails >= 4. Conditional tested conditions are bad: it is like you are testing a different behavior, while it is the same.

Reply:
I agree. But my plan is to deprecate Rails 3 compatibility right after next release (after merging this PR) which will remove one branch of this condition.